### PR TITLE
[JBIDE-22647] Code updates to support changes from [OSJC-261]

### PR DESCRIPTION
**Pull request by jcantrill**
cc @fbricon @adietish

### CODE REVIEW

**Comment by jeffmaury** [comment]
Got exception while trying to deploy nodejs-example:

```
!ENTRY org.jboss.tools.openshift.ui 4 -1 2016-06-27 10:33:38.906
!MESSAGE Template in version "undefined" cannot be handled as a Template: no kind "Template" is registered for version "undefined"
!STACK 0
com.openshift.restclient.BadRequestException: https://10.1.2.2:8443/oapi/v1/namespaces/test/processedtemplates Template in version "undefined" cannot be handled as a Template: no kind "Template" is registered for version "undefined"
	at com.openshift.internal.restclient.okhttp.ResponseCodeInterceptor.createOpenShiftException(ResponseCodeInterceptor.java:96)
	at com.openshift.internal.restclient.okhttp.ResponseCodeInterceptor.intercept(ResponseCodeInterceptor.java:59)
	at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:190)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:163)
	at okhttp3.RealCall.execute(RealCall.java:57)
	at com.openshift.internal.restclient.DefaultClient.execute(DefaultClient.java:208)
	at com.openshift.internal.restclient.DefaultClient.create(DefaultClient.java:170)
	at com.openshift.internal.restclient.capability.server.ServerTemplateProcessing.process(ServerTemplateProcessing.java:44)
	at com.openshift.internal.restclient.capability.resources.ProjectTemplateProcessing.process(ProjectTemplateProcessing.java:52)
	at org.jboss.tools.openshift.internal.ui.job.CreateApplicationFromTemplateJob$1.visit(CreateApplicationFromTemplateJob.java:69)
	at org.jboss.tools.openshift.internal.ui.job.CreateApplicationFromTemplateJob$1.visit(CreateApplicationFromTemplateJob.java:1)
	at com.openshift.internal.restclient.model.KubernetesResource.accept(KubernetesResource.java:91)
	at org.jboss.tools.openshift.internal.ui.job.CreateApplicationFromTemplateJob.doRun(CreateApplicationFromTemplateJob.java:63)
	at org.jboss.tools.openshift.internal.common.core.job.AbstractDelegatingMonitorJob.run(AbstractDelegatingMonitorJob.java:37)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)

```

from the UI:

image

**Comment by jcantrill** [comment]
@jeffmaury did you install the corresponding changes from the rest client?

**Comment by jeffmaury** [comment]
Yes if you look the class path, you'll find some okhttp classes.
I did:
pull you PR on openshift-rest-client then mvn clean install
pull your PR on jbosstools-openshift then mvn process-resources (to get the libs) 
then work as usual in Eclipse

**Comment by jeffmaury** [comment]
The test was used against CDK 2.1RC5 (http://cdk-builds.usersys.redhat.com/builds/nightly/16-Jun-2016/)

**Comment by jeffmaury** [comment]
Fix: the exact URLCDK2.1RC5 is http://cdk-builds.usersys.redhat.com/builds/weekly/16-Jun-2016.rc5/

**Comment by jcantrill** [comment]
cc @fbricon @jeffmaury @adietish updated after rebase

**Comment by jeffmaury** [comment]
Test is now ok with ca3d285 commit

**Comment by jcantrill** [comment]
@fbricon @jeffmaury @adietish @scabanovich  any chance of getting some traction on this PR with an LGTM so it can be merged.  I would like our outstanding PRs to start basing off these changes

**Comment by jeffmaury** [comment]
@jcantrill I would like all of us to switch to this new implementation but this is something I planned to discuss during yesterday's Assparta call but you didn't show up as I'm not clear on the status of the changes of the Origin Rest client (seems merged to master but it is deployed on Nexus ?). On my side, the problem I found is fixed so I would favor +1 for the merge (cc @alexeykazakov )

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/connection/Connection.java [hunk 1]
```diff
@@ -246,18 +243,14 @@ public boolean connect() throws OpenShiftException {
...
+			if (!context.isAuthorized() 
+					&& credentialsPrompter != null
+					&& promptCredentialsEnabled){
+				credentialsPrompter.promptAndAuthenticate(this, null);
 			} else {
-				if (isEnablePromptCredentials()
-						&& credentialsPrompter != null) {
-					credentialsPrompter.promptAndAuthenticate(this, null);
-				}
+				updateCredentials(context);
```
Comment (scabanovich) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/connection/Connection.java]: If context is not authorized and prompter is null (or not enabled), updateCredentials will be called anyway. Is it as it should be?

**Comment by fbricon** [comment]
After rebasing the code to latest master, I got this weird behavior where login doesn't work OOTB, even though credentials are properly set and manual logging works fine: http://screencast.com/t/hGNYeyWm0sHb

File: plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/cdk/server/core/internal/adapter/VagrantPoller.java [hunk 2]
```diff
@@ -187,28 +189,22 @@ private boolean checkOpenShiftHealth(IServer server, int timeout) throws OpenShi
 		if( adb == null ) {
 			return false;
 		}
-		String url = adb.getOpenShiftHost() + ":" + adb.getOpenShiftPort() + "/healthz/ready";
+		String url = adb.getOpenShiftHost() + ":" + adb.getOpenShiftPort();
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/cdk/server/core/internal/adapter/VagrantPoller.java]: why removing the "/healthz/ready" part?

**Comment by adietish** [comment]
The latest content of this PR doesnt match the latest client lib API any more. One needs to change ApplicationSourceFromTemplateModel:
{code:title=ApplicationSourceFromTemplateModel#updateTemplateParameters}
setItems(template.getItems());
{code}
to
{code:title=ApplicationSourceFromTemplateModel#updateTemplateParameters}
setItems(template.getObjects());
{code}
@jcantrill can you please apply this change on your PR so that rebasing gets easier?

**Comment by adietish** [comment]
@jcantrill after updating my branch merging works fine. Sorry for the inconvenience, no further changes required.

**Comment by adietish** [comment]
@fbricon @jcantrill I can see the same bug: the connection dialog appears even though the password is available from secure storage.

**Comment by jcantrill** [comment]
@adietish @fbricon  fixed the username/token/password prompt and loading issue as well as rebased on head.  Please test/review/etc for merge.

File: plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/cdk/server/core/internal/adapter/VagrantPoller.java (hunk 2 above)
Comment (jcantrill) [inline plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/cdk/server/core/internal/adapter/VagrantPoller.java]: Functionality moved into the client and is essentially here: https://github.com/jbosstools/jbosstools-openshift/pull/1239/commits/a021ad2ba8f1ce6f06bbfa452837b6104831750e#diff-29e83e4203e38d1faadbabcea0001398R204

**Comment by jcantrill** [comment]
@adietish @fbricon ok... its really been updated to fix the compile and test failures.

**Comment by fbricon** [comment]
So the silent connection now works. Scaling pods, Port forwarding, Pod logs, Resource Editing work

Deleting a project however generates a s*ton of errors in the error log, from all the okhttp threads running (watchers):

```
Exception in thread "OkHttp Dispatcher" com.openshift.restclient.authorization.ResourceForbiddenException: User "openshift-dev" cannot list pods in project "sample-project" User "openshift-dev" cannot list pods in project "sample-project"
	at com.openshift.internal.restclient.okhttp.ResponseCodeInterceptor.createOpenShiftException(ResponseCodeInterceptor.java:99)
	at com.openshift.internal.restclient.okhttp.ResponseCodeInterceptor.intercept(ResponseCodeInterceptor.java:59)
	at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:190)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:163)
	at okhttp3.RealCall.access$100(RealCall.java:30)
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:127)
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

I also get a bunch of:
```
juil. 22, 2016 11:21:16 PM okhttp3.internal.Platform log
AVERTISSEMENT: A connection to https://10.1.2.2:8443/ was leaked. Did you forget to close a response body?
juil. 22, 2016 11:21:16 PM okhttp3.internal.Platform log
AVERTISSEMENT: A connection to https://10.1.2.2:8443/ was leaked. Did you forget to close a response body?
juil. 22, 2016 11:21:16 PM okhttp3.internal.Platform log
AVERTISSEMENT: A connection to https://10.1.2.2:8443/ was leaked. Did you forget to close a response body?
juil. 22, 2016 11:21:16 PM okhttp3.internal.Platform log
AVERTISSEMENT: A connection to https://10.1.2.2:8443/ was leaked. Did you forget to close a response body?
juil. 22, 2016 11:21:16 PM okhttp3.internal.Platform log
AVERTISSEMENT: A connection to https://10.1.2.2:8443/ was leaked. Did you forget to close a response body?
juil. 22, 2016 11:21:16 PM okhttp3.internal.Platform log
AVERTISSEMENT: A connection to https://10.1.2.2:8443/ was leaked. Did you forget to close a response body?
juil. 22, 2016 11:21:16 PM okhttp3.internal.Platform log
AVERTISSEMENT: A connection to https://10.1.2.2:8443/ was leaked. Did you forget to close a response body?
juil. 22, 2016 11:21:16 PM okhttp3.internal.Platform log
AVERTISSEMENT: A connection to https://10.1.2.2:8443/ was leaked. Did you forget to close a response body?
```

Eventually, once the project is deleted, I still see some dangling okhttp threads in the jvm

**Comment by jcantrill** [comment]
@fbricon let me redeploy the rest client.  I'm wondering if https://github.com/openshift/openshift-restclient-java/pull/188 fixed it.

**Comment by jeffmaury** [comment]
With the latest SNAPSHOT it fixed the project deletion example.
However, I have been able to get an empty set of resources returned when an application is created:
[2016-07-25 at 12-41-50.zip](https://github.com/jbosstools/jbosstools-openshift/files/381281/2016-07-25.at.12-41-50.zip)

**Comment by fbricon** [comment]
Using the nodejs builder, we should get 5 resources: imagestream, service, buildconfig, deploymentconfig, route.

**Comment by jcantrill** [comment]
@fbricon are you able to confirm this issue?  Trying with the JBT code I have results in:

image

**Comment by jcantrill** [comment]
scrach that...figured out i was running on wrong branch.  verifying....

**Comment by jcantrill** [comment]
@fbricon @jeffmaury I tried on the right branch and got:
image

Anything else you can point me to?

**Comment by fbricon** [comment]
@jcantrill @jeffmaury resources are created for me on this branch. @jeffmaury have you run a mvn clean build before testing?

**Comment by fbricon** [comment]
so deleting a project works fine now. However, I'm still seeing 9 okhttp threads after the deletion. Is it normal? Is this something covered by PR #1244?

**Comment by jeffmaury** [comment]
The problem with no resources being created is not a permanent state. In order to reproduce, repeat the following steps in order:

1. Create an Openshift project
2. Deploy the nodejs:0.10 (builder) into that project
3. Delete the Openshift project
4. Goto 1.

**Comment by fbricon** [comment]
@jeffmaury I can't reproduce your problem

**Comment by jeffmaury** [comment]
@fbricon I reproduced it with a fresh CDK VM at the third try

**Comment by jcantrill** [comment]
@fbricon alternative issue I'm seeing is 2 connections to the same server and project, I am not seeing updates.  I'm wondering if this would be fixed by #1244 which includes a connection in the map.  I'm THINKING there is funny business that PR may resolve.

In regards to the thread count, we may need to tweak how threads are created.  It's worth discussing further as I had to make changes in the client which could probably be optimized.

**Comment by jcantrill** [comment]
merging per conv with @fbricon.  Outstanding issues to be fixed in subsequent PRs

**Pull request merged**